src/Operations.jl

```diff
@@ -961,21 +983,57 @@ function assert_can_add(ctx::Context, pkgs::Vector{PackageSpec})
     end
 end
 
+function tiered_resolve(ctx::Context, pkgs::Vector{PackageSpec})
+    try # do not modify existing subgraph
+        printpkgstyle(ctx, :Resolve_Tier, "All dependency versions.")
+        return targeted_resolve(ctx, pkgs, PRESERVE_ALL)
+    catch err
+        err isa ResolverError || rethrow()
+    end
+    try # do not modify existing direct deps
+        printpkgstyle(ctx, :Resolve_Tier, "Direct dependency versions.")
+        return targeted_resolve(ctx, pkgs, PRESERVE_DIRECT)
+    catch err
+        err isa ResolverError || rethrow()
+    end
+    printpkgstyle(ctx, :Resolve_Tier, "Direct dependency semver constraints.")
+    return targeted_resolve(ctx, pkgs, PRESERVE_SEMVER)
```

Comment:
Shouldn't there be a call to `targeted_resolve` with `PRESERVE_NONE` here?

Reply:
The idea is that regular operation shouldn't silently break semver. It would be weird if I did `add Foo` and now my code won't run. From manual testing, it seems `PRESERVE_SEMVER` is actually surprisingly flexible. If the user decides they don't care, they can always just do `add --preserve=none Foo`

Reply:
I mean we have the `[compat]` section there for a reason though. And we did exactly have the `PRESERVE_SEMVER` strategy before and it leads to a lot of resolver errors in situations where a tries the following:

Add A that installs e.g. version 0.8
Add B that has compat `0.7` with A.

This gives a resolver error because A is not allowed to downgrade from 0.8. If the user had instead installed these in the opposite order, everything would be fine. Adding both packages at the same time would also work ok.

Reply:
Fair point about `compat`. I'll change this as well

Reply:
Here is an example in the wild: https://github.com/JuliaStats/StatsKit.jl/issues/20.